File: ts-src/src/providers/codex.ts
import GLib from 'gi://GLib';
import Gio from 'gi://Gio';
import { ProviderData, PROVIDER_IDS } from '../types.js';

function directoryExists(path: string): boolean {
  try {
    const file = Gio.File.new_for_path(path);
    return file.query_exists(null);
  } catch {
    return false;
  }
}

function readJsonFile(path: string): any {
  try {
    const file = Gio.File.new_for_path(path);
    if (!file.query_exists(null)) return null;
    const [ok, contents] = file.load_contents(null);
    if (!ok) return null;
    const decoder = new TextDecoder('utf-8');
    return JSON.parse(decoder.decode(contents));
  } catch {
    return null;
  }
}

export function fetchCodexData(): ProviderData {
  const homeDir = GLib.get_home_dir();
  const codexDir = GLib.build_filename(homeDir, '.codex');

  if (!directoryExists(codexDir)) {
    return {
      id: PROVIDER_IDS.CODEX,
      name: 'Codex',
      usedTokens: 0,
      limitTokens: 100,
      usedCredits: null,
      limitCredits: null,
      resetAt: null,
      error: 'Not installed',
    };
  }

  // Try common Codex usage file locations
  const candidates = [
    GLib.build_filename(codexDir, 'usage.json'),
    GLib.build_filename(codexDir, 'state.json'),
    GLib.build_filename(codexDir, 'config.json'),
  ];

  for (const path of candidates) {
    try {
      const data = readJsonFile(path);
      if (!data) continue;

      // Adapt to whatever schema Codex uses
      if (typeof data.usage_pct === 'number') {
        return {
          id: PROVIDER_IDS.CODEX,
          name: 'Codex',
          usedTokens: Math.round(data.usage_pct),
          limitTokens: 100,
          usedCredits: data.credits_used ?? null,
          limitCredits: data.credits_limit ?? null,
          resetAt: data.resets_at ? new Date(data.resets_at) : null,
          error: null,
        };
      }

      if (typeof data.tokens_used === 'number' && typeof data.tokens_limit === 'number') {
        const pct = Math.round((data.tokens_used / data.tokens_limit) * 100);
        return {
          id: PROVIDER_IDS.CODEX,
          name: 'Codex',
          usedTokens: pct,
          limitTokens: 100,
          usedCredits: null,
          limitCredits: null,
          resetAt: data.reset_at ? new Date(data.reset_at) : null,
          error: null,
        };
      }
    } catch {
      // skip malformed files
    }
  }

  // .codex dir exists but no usage data found
  return {
    id: PROVIDER_IDS.CODEX,
    name: 'Codex',
    usedTokens: 0,
    limitTokens: 100,
    usedCredits: null,
    limitCredits: null,
    resetAt: null,
    error: 'No usage data',
  };
}

File: src/providers/codex.rs
use serde::Deserialize;
use std::path::PathBuf;
use super::ProviderData;
use crate::config::home_dir;

#[derive(Debug, Deserialize)]
struct CodexUsage {
    usage_pct: Option<f64>,
    tokens_used: Option<u64>,
    tokens_limit: Option<u64>,
    credits_used: Option<f64>,
    credits_limit: Option<f64>,
    resets_at: Option<String>,
    reset_at: Option<String>,
}

fn codex_dir() -> PathBuf {
    home_dir().join(".codex")
}

fn error_entry(msg: &str) -> ProviderData {
    super::error_entry("codex", "Codex", msg)
}

pub fn fetch_codex() -> ProviderData {
    let dir = codex_dir();
    if !dir.exists() {
        return error_entry("Not installed");
    }

    let candidates = ["usage.json", "state.json", "config.json"];
    for name in &candidates {
        let path = dir.join(name);
        if !path.exists() {
            continue;
        }
        let Ok(contents) = std::fs::read_to_string(&path) else { continue };
        let Ok(data) = serde_json::from_str::<CodexUsage>(&contents) else { continue };

        if let Some(pct) = data.usage_pct {
            let reset_at = data.resets_at.as_deref()
                .and_then(|s| s.parse().ok());
            return ProviderData {
                id: "codex".to_string(),
                name: "Codex".to_string(),
                utilization: pct as f32,
                reset_at,
                used_credits: data.credits_used,
                limit_credits: data.credits_limit,
                ..Default::default()
            };
        }

        if let (Some(used), Some(limit)) = (data.tokens_used, data.tokens_limit) {
            let utilization = if limit > 0 {
                (used as f64 / limit as f64 * 100.0) as f32
            } else {
                0.0
            };
            let reset_at = data.reset_at.as_deref()
                .and_then(|s| s.parse().ok());
            return ProviderData {
                id: "codex".to_string(),
                name: "Codex".to_string(),
                utilization,
                reset_at,
                tokens_used: Some(used),
                ..Default::default()
            };
        }
    }

    error_entry("No usage data")
}
